src/middleware/io/can_interface.h

```diff
@@ -0,0 +1,183 @@
+    void open_socket() override;
+
+    void data_rec(struct can_frame& receive_frame_,
+                  boost::asio::posix::basic_stream_descriptor<>& stream);
+
+  private:
+    struct can_frame receive_frame_;
+};
+} // namespace io
+} // namespace middleware
+} // namespace goby
+
+template <const goby::middleware::Group& line_in_group,
+          const goby::middleware::Group& line_out_group,
+          goby::middleware::io::PubSubLayer publish_layer,
+          goby::middleware::io::PubSubLayer subscribe_layer>
+void goby::middleware::io::CanThread<line_in_group, line_out_group, publish_layer,
+                                     subscribe_layer>::open_socket()
+{
+    int can_socket;
+
+    struct sockaddr_can addr_;
+    struct can_frame receive_frame_;
+    struct ifreq ifr_;
+    can_socket = socket(PF_CAN, SOCK_RAW, CAN_RAW);
+
+    std::vector<struct can_filter> filters;
+
+    for (auto x : this->cfg().filter()) { filters.push_back({x.can_id(), x.can_mask()}); }
+
+    if (filters.size())
+    {
+        setsockopt(can_socket, SOL_CAN_RAW, CAN_RAW_FILTER, filters.data(),
+                   sizeof(can_filter) * filters.size());
+    }
+    std::strcpy(ifr_.ifr_name, this->cfg().interface().c_str());
+
+    ioctl(can_socket, SIOCGIFINDEX, &ifr_);
+
+    addr_.can_family = AF_CAN;
+    addr_.can_ifindex = ifr_.ifr_ifindex;
+    if (bind(can_socket, (struct sockaddr*)&addr_, sizeof(addr_)) < 0)
+    {
+        glog.is_die() && glog << "Error in socket bind to interface " << this->cfg().interface()
+                              << ": " << std::strerror(errno) << std::endl;
+    }
+
+    this->mutable_socket().assign(can_socket);
+}
+
+template <const goby::middleware::Group& line_in_group,
+          const goby::middleware::Group& line_out_group,
+          goby::middleware::io::PubSubLayer publish_layer,
+          goby::middleware::io::PubSubLayer subscribe_layer>
+void goby::middleware::io::CanThread<line_in_group, line_out_group, publish_layer,
+                                     subscribe_layer>::async_write(const std::string& bytes)
+{
+    // TODO: Check frame validity?
+
+    this->mutable_socket().async_write_some(
+        boost::asio::buffer(bytes),
+        [this](const boost::system::error_code& ec, std::size_t bytes_transferred) {
+            if (!ec && bytes_transferred > 0)
+            {
+                this->handle_write_success(bytes_transferred);
+            }
+            else
+            {
+                this->handle_write_error(ec);
+            }
+        });
+}
+
+template <const goby::middleware::Group& line_in_group,
+          const goby::middleware::Group& line_out_group,
+          goby::middleware::io::PubSubLayer publish_layer,
+          goby::middleware::io::PubSubLayer subscribe_layer>
+void goby::middleware::io::CanThread<line_in_group, line_out_group, publish_layer,
+                                     subscribe_layer>::async_read()
+{
+    this->mutable_socket().async_read_some(
```

Comment:
Are you sure async_read_some will always work? From https://www.boost.org/doc/libs/1_73_0/doc/html/boost_asio/reference/basic_stream_socket/async_read_some.html: "The read operation may not read all of the requested number of bytes. Consider using the async_read function if you need to ensure that the requested amount of data is read before the asynchronous operation completes. "

Reply:
Should this be `async_read` instead?

Reply:
I think you are right. I'll push an update to both places where I use this.